pkg/alertmanager/alertmanagerdiscovery/dns.go

```diff
@@ -0,0 +1,126 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package alertmanagerdiscovery
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/grafana/dskit/cache"
+	"github.com/grafana/dskit/dns"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/discovery"
+	"github.com/prometheus/prometheus/discovery/refresh"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+
+	util_log "github.com/grafana/mimir/pkg/util/log"
+)
+
+const (
+	mechanismName = "dns_sd"
+)
+
+// NewDNSProvider returns a new dns.Provider that can be used to build discovery.Config instances.
+func NewDNSProvider(reg prometheus.Registerer) *dns.Provider {
+	// We need to prefix and add a label to the metrics for the DNS resolver because, unlike other mimir components,
+	// it doesn't already have the `cortex_` prefix and the `component` label to the metrics it emits
+	dnsProviderReg := prometheus.WrapRegistererWithPrefix(
+		"cortex_",
+		prometheus.WrapRegistererWith(
+			prometheus.Labels{"component": "ruler"},
+			reg,
+		),
+	)
+
+	return dns.NewProvider(util_log.Logger, dnsProviderReg, dns.GolangResolverType)
+}
+
+// BuildDiscoveryConfigs populates a map of alert manager URLs to discovery.Config instances.
+func BuildDiscoveryConfigs(alertManagerUrls string, discoveryConfigs map[string]discovery.Config, alertmanagerRefreshInterval time.Duration, resolver *dns.Provider) error {
```

Comment:
Design wise, I think that it's harder to follow code where the output value is passed in input. I think you can just return `map[string]discovery.Config` from this function.